File: Services/pkg/permission/repository.go
package permission

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	txcontext "github.com/ChangerzaryX1602/SkillSync/pkg/tx_context"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"

	"github.com/gofiber/fiber/v2"
	helpers "github.com/zercle/gofiber-helpers"
	"gorm.io/gorm"
)

type permissionRepository struct {
	resources models.Resources
}

func NewPermissionRepository(resources models.Resources) domain.PermissionRepository {
	return &permissionRepository{resources: resources}
}

func (r *permissionRepository) Migrate() error {
	return r.resources.MainDbConn.AutoMigrate(&models.Permission{})
}

func (r *permissionRepository) CreatePermission(ctx context.Context, permission models.Permission) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Create(&permission).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *permissionRepository) GetPermission(ctx context.Context, id uint) (*models.Permission, *helpers.ResponseError) {
	db := r.getDB(ctx)
	var permission models.Permission
	if err := db.WithContext(ctx).Where("id = ?", id).First(&permission).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return &permission, nil
}

func (r *permissionRepository) GetPermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Permission, *models.Pagination, *models.Search, *helpers.ResponseError) {
	db := r.getDB(ctx)
	var permissions []models.Permission
	db = db.WithContext(ctx).Model(&models.Permission{})
	db = utils.ApplySearch(ctx, r.resources.FastHTTPClient, db, search, false)
	db = utils.ApplyPagination(db, &pagination, models.Permission{})
	if err := db.Find(&permissions).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil, nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, nil, nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return permissions, &pagination, &search, nil
}

func (r *permissionRepository) UpdatePermission(ctx context.Context, id uint, permission models.Permission) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Where("id = ?", id).First(&models.Permission{}).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := db.WithContext(ctx).Model(&models.Permission{}).Where("id = ?", id).Updates(permission).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *permissionRepository) DeletePermission(ctx context.Context, id uint) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Where("id = ?", id).First(&models.Permission{}).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := db.WithContext(ctx).Delete(&models.Permission{}, id).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *permissionRepository) getDB(ctx context.Context) *gorm.DB {
	if tx := txcontext.GetTxFromContext(ctx); tx != nil {
		return tx
	}
	return r.resources.MainDbConn
}

File: hono-clean-architecture/pkg/permission/repository.ts
import { NodePgDatabase } from "drizzle-orm/node-postgres";
import { eq, and, sql, count } from "drizzle-orm";
import { permissions, Permission, Pagination, Search, Resources } from "../models";
import { PermissionRepository as IPermissionRepository } from "../domain";
import { ResponseError, createError, whereAmI } from "../../internal/infrastructure/custom_error";
import { applySearch, getOffset, getLimit } from "../utils";
import { PermissionCache, newPermissionCache } from "./cache";

export class PermissionRepository implements IPermissionRepository {
  private db: NodePgDatabase;
  private cache: PermissionCache;

  constructor(resources: Resources) {
    this.db = resources.mainDbConn;
    this.cache = newPermissionCache(resources.redisStorage);
  }

  async migrate(): Promise<void> {
    console.log("Permission table migration handled by Drizzle Kit");
  }

  async createPermission(permissionData: {
    group: string;
    name: string;
  }): Promise<{ permission: Permission | null; error: ResponseError | null }> {
    const existing = await this.db
      .select()
      .from(permissions)
      .where(
        and(eq(permissions.group, permissionData.group), eq(permissions.name, permissionData.name))
      )
      .limit(1);

    if (existing.length > 0) {
      return {
        permission: null,
        error: createError(409, whereAmI(), "Conflict", "Permission already exists."),
      };
    }

    try {
      const result = await this.db
        .insert(permissions)
        .values({ group: permissionData.group, name: permissionData.name })
        .returning();

      await this.cache.invalidateAllLists();
      return { permission: result[0], error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        permission: null,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async getPermission(
    id: number
  ): Promise<{ permission: Permission | null; error: ResponseError | null }> {
    const cached = await this.cache.get(id);
    if (cached) {
      return { permission: cached, error: null };
    }

    try {
      const result = await this.db
        .select()
        .from(permissions)
        .where(eq(permissions.id, id))
        .limit(1);

      if (result.length === 0) {
        return {
          permission: null,
          error: createError(404, whereAmI(), "Not Found", "Permission not found"),
        };
      }

      await this.cache.set(result[0]);
      return { permission: result[0], error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        permission: null,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async getPermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    permissions: Permission[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }> {
    try {
      const searchCondition = applySearch(search, [permissions.group, permissions.name]);

      const countQuery = searchCondition
        ? this.db.select({ count: count() }).from(permissions).where(searchCondition)
        : this.db.select({ count: count() }).from(permissions);

      const countResult = await countQuery;
      const total = Number(countResult[0]?.count ?? 0);

      let query = this.db.select().from(permissions);

      if (searchCondition) {
        query = query.where(searchCondition) as typeof query;
      }

      if (pagination.orderBy) {
        const orderDirection = pagination.order === "desc" ? sql`DESC` : sql`ASC`;
        query = query.orderBy(
          sql`${sql.identifier(pagination.orderBy)} ${orderDirection}`
        ) as typeof query;
      }

      const result = await query.limit(getLimit(pagination)).offset(getOffset(pagination));

      return {
        permissions: result,
        pagination: { ...pagination, total },
        search,
        error: null,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        permissions: [],
        pagination,
        search,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async updatePermission(
    id: number,
    permissionData: Partial<Permission>
  ): Promise<ResponseError | null> {
    try {
      const current = await this.db
        .select()
        .from(permissions)
        .where(eq(permissions.id, id))
        .limit(1);

      if (current.length === 0) {
        return createError(404, whereAmI(), "Not Found", "Permission not found");
      }

      await this.db
        .update(permissions)
        .set({ ...permissionData, updatedAt: new Date() })
        .where(eq(permissions.id, id));

      await this.cache.invalidate(id);
      await this.cache.invalidateAllLists();

      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async deletePermission(id: number): Promise<ResponseError | null> {
    try {
      const permission = await this.db
        .select()
        .from(permissions)
        .where(eq(permissions.id, id))
        .limit(1);

      if (permission.length === 0) {
        return createError(404, whereAmI(), "Not Found", "Permission not found");
      }

      await this.db.delete(permissions).where(eq(permissions.id, id));
      await this.cache.invalidate(id);
      await this.cache.invalidateAllLists();

      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async getPermissionByGroupAndName(
    group: string,
    name: string
  ): Promise<{ permission: Permission | null; error: ResponseError | null }> {
    try {
      const result = await this.db
        .select()
        .from(permissions)
        .where(and(eq(permissions.group, group), eq(permissions.name, name)))
        .limit(1);

      if (result.length === 0) {
        return {
          permission: null,
          error: createError(404, whereAmI(), "Not Found", "Permission not found"),
        };
      }

      return { permission: result[0], error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        permission: null,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }
}

export function newPermissionRepository(resources: Resources): IPermissionRepository {
  return new PermissionRepository(resources);
}
